ext/compressors/lz4/lz4_compress.c

```diff
@@ -121,40 +138,143 @@ lz4_decompress(WT_COMPRESSOR *compressor, WT_SESSION *session,
     size_t *result_lenp)
 {
 	WT_EXTENSION_API *wt_api;
-	char *compressed_data;
+	LZ4_PREFIX prefix;
 	int decoded;
-	size_t src_data_len;
+	uint8_t *dst_tmp;
+
+	(void)src_len;					/* Unused parameters */
 
 	wt_api = ((LZ4_COMPRESSOR *)compressor)->wt_api;
 
-	/* Retrieve compressed length from start of the data buffer. */
-	src_data_len = *(size_t *)src;
-	if (src_data_len + sizeof(size_t) > src_len) {
-		(void)wt_api->err_printf(wt_api,
-		    session,
-		    "lz4_decompress: stored size exceeds buffer size");
-		return (WT_ERROR);
+	/*
+	 * Retrieve the true length of the compressed block and source and the
+	 * decompressed bytes to return from the start of the source buffer.
+	 */
+	memcpy(&prefix, src, sizeof(LZ4_PREFIX));
+
+	/*
+	 * Decompress, starting after the prefix bytes. Use safe decompression:
+	 * we rely on decompression to detect corruption.
+	 *
+	 * Two code paths, one with and one without a bounce buffer. When doing
+	 * raw compression, we compress to a target size irrespective of row
+	 * boundaries, and return to our caller a "useful" compression length
+	 * based on the last complete row that was compressed. Our caller stores
+	 * that length, not the length of bytes actually compressed by LZ4. In
+	 * other words, our caller doesn't know how many bytes will result from
+	 * decompression, likely hasn't provided us a large enough buffer, and
+	 * we have to allocate a scratch buffer.
+	 */
```

Comment:
> 1. Decompress up to a maximum number of bytes, knowing compress size (your request).

As you noted, our issue is a buffer that's too small to hold the fully uncompressed block, so this doesn't help.

> 2. Evaluate the maximum nb of bytes that can be compressed into a fixed destination size, then use this information to size down this amount to the last full-size record and then request LZ4 to compress that amount.

This will be slower than the current approach (because we're doing two compression calls), so I think we'd rather stay with what we have.

> 3. Decompress up to a maximum number of bytes, without knowing compressed size (but knowing at least the maximum possible compressed size).

Understood, presumably this would be a new function?

Is there any possibility of a unique error return from `LZ4_decompress_safe`? If there was one known error that meant "end of the user buffer", I think that would be sufficient, we could treat that error as success.

Otherwise, I suspect what we have now is the best we can do, and we'll have to investigate ways to change our layering so additional information is available when we're decompressing, and we can allocate the right buffer length.